### ISSUE TITLE
[Feature Request] PostgreSQL: Unique Sequence for each entity

### ISSUE THREAD

**Issue by amatosg** [jhipster/generator-jhipster#9322]
##### **Overview of the feature request**

I see no reason for having a single sequence for **all** the entities. It would make more sense (to me, at least) if each entity has its own sequence.

##### **Motivation for or Use Case**

Entities' id would have a logic sequence. For me is *weird* to see one entity having id 1050, 1055 and then 1065, while others could have 1056, 1057.

##### **Related issues or PR**

I've searched both open and closed issues and haven't found anything.

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by jdubois** [comment]
I believe that what we do is the default in Hibernate, so that would be against [policy 1](https://www.jhipster.tech/policies/), could you check?

On a personal side, I'd rather have just one sequence, I find this simpler.

**Comment by cbornet** [comment]
We use the sequence generator with Hilo algorithm which allows performant batch. See https://vladmihalcea.com/hibernate-identity-sequence-and-table-sequence-generator/

**Comment by jdubois** [comment]
Yes I think we have the recommended approach from Vlad, so I'm closing this (of course you can still prove me wrong and I'll re-open)

**Comment by amatosg** [comment]
What about using `id BIGINT GENERATED ALWAYS AS IDENTITY`? That way each table would have its own sequence generator (just like in MySQL/MariaDB). With that, things get simplify because:

1. There is no need to create sequences
2. The IDs for each entity will be correlative
3. No additional queries are needed as there are no sequences 

My 3rd point would make it even more performant than the Hilo algorithm as there are no queries involved for generating the id.

The only drawback I see is that that feature was introduced in PostgreSQL 10 in October 2017, which it might impact on users having 9.x.

What do you guys think?

**Comment by gmarziou** [comment]
> 3\. No additional queries are needed as there are no sequences

Have you checked this with datasource-proxy or p6spy?
I don't understand how Hibernate would know the id without re-loading the entity.

Using sequences in Hibernate is better because it can provision a batch of new ids in one statement and then use them for many inserts without having to ask sequence fir each insert.

That said, I'd prefer too to have a sequence per table.

**Comment by murdos** [comment]
> 
> That said, I'd prefer too to have a sequence per table.

I prefer that too.
Would it be ok if it was available as an option?

**Comment by amatosg** [comment]
> Using sequences in Hibernate is better because it can provision a batch of new ids in one statement and then use them for many inserts without having to ask sequence for each insert.

But not having sequences means there is no need to ask for next id(s). Those values are generated just before commiting (AFAIK).

> I don't understand how Hibernate would know the id without re-loading the entity

Hibernate doesn't have to know the ID. On `save(T)` hibernate returns the entity with the generated ID.

> Would it be ok if it was available as an option?

That would be nice if it's not accepted as default. I think I have the code already done somewhere.

**Comment by jdubois** [comment]
I really like the option from @amatosg then we can't have this right now as it's only on Postgresql 10. As far as I understand it will be available on GCP in the 2nd quarter of 2019. So we can't rely on a feature that's not available on all major cloud providers.

So let's wait a few months, maybe 1 year, and then we can do this.

**Comment by amatosg** [comment]
Another way to give each entity a sequence, is using the `SMALLSERIAL`/`SERIAL`/`BIGSERIAL` pseudo-type. As an example, [this statement](https://www.postgresql.org/docs/9.1/datatype-numeric.html#DATATYPE-SERIAL):
``` SQL
CREATE TABLE table_name(
    id SERIAL
);
```
Will be translated into:
``` SQL
CREATE SEQUENCE table_name_id_seq;
 
CREATE TABLE table_name (
    id integer NOT NULL DEFAULT nextval('table_name_id_seq')
);
 
ALTER SEQUENCE table_name_id_seq
OWNED BY table_name.id;
```

And GCP's documentation [states](https://cloud.google.com/sql/docs/postgres/quickstart#create_a_database_and_upload_data):
``` SQL
CREATE TABLE guestbook (guestName VARCHAR(255), content VARCHAR(255), entryID SERIAL PRIMARY KEY);

I think it's save to use this approach, what do you guys think?

**Comment by jdubois** [comment]
@amatosg `SERIAL` doesn't look to be such a great idea for performance: https://vladmihalcea.com/postgresql-serial-column-hibernate-identity/

**Comment by pvliss** [comment]
@jdubois I am also generally in favor of using either sequence or identity generation strategies as well. Actually, sequences are the most recommended way according to @vladmihalcea(https://vladmihalcea.com/hibernate-identity-sequence-and-table-sequence-generator/ ), at least in terms of performance and are supported by all relational DBs AFAIK. I could create a PR if you are interested

**Comment by jdubois** [comment]
At the moment we have auto numbers with databases that support it, and sequences for the others. I believe it's the best solutions for now.

**Comment by pvliss** [comment]
Oh I just had a look at the code and indeed that is already the case. Thought it was TABLE generation by default. Nothing to do there . You are right. Sorry for wasting your time. Got to get some rest...

**Comment by jdubois** [comment]
No problem @pvliss I know (and appreciate!) that you are trying to help

**Comment by cklee75** [comment]
Seems there are more benefits to use unique sequence for each entity than global sequence. May refer to [Which is better: single global sequence vs. sequence per table?](https://stackoverflow.com/questions/7774644/which-is-better-single-global-sequence-vs-sequence-per-table) and [Asking for opinions : One sequence for all tables
](https://stackoverflow.com/questions/1536479/asking-for-opinions-one-sequence-for-all-tables). Posts are for Oracle, but the same should apply to other databases that support sequence.

**Comment by laariche-khalid** [comment]
> I really like the option from @amatosg then we can't have this right now as it's only on Postgresql 10. As far as I understand it will be available on GCP in the 2nd quarter of 2019. So we can't rely on a feature that's not available on all major cloud providers.
> 
> So let's wait a few months, maybe 1 year, and then we can do this.

Hi , is there any news on this feature?

Merci pour votre feedback :)

**Comment by lanied** [comment]
It seems to have been added in version 8 through the custom field @SequenceGeneratorName 
example,
/*
 * Integer custom id and sequence
 */
@ChangelogDate(20200805000200)
entity EntityCustomSequence {
  @SequenceGeneratorName(entityIntegerIdSeq)
  id Long
  name String
}

**Comment by lanied** [comment]
> > I really like the option from @amatosg then we can't have this right now as it's only on Postgresql 10. As far as I understand it will be available on GCP in the 2nd quarter of 2019. So we can't rely on a feature that's not available on all major cloud providers.
> > So let's wait a few months, maybe 1 year, and then we can do this.
> 
> Hi , is there any news on this feature?
> 
> Merci pour votre feedback :)

It seems to have been added in version 8 through the custom field @SequenceGeneratorName 
example,
/*
 * Integer custom id and sequence
 */
@ChangelogDate(20200805000200)
entity EntityCustomSequence {
  @SequenceGeneratorName(entityIntegerIdSeq)
  id Long
  name String
}

**Comment by yelhouti** [comment]
@jdubois @mshima could you please reopon this, as recommendation changed for hibernate 6
https://vladmihalcea.com/hibernate-default-entity-sequence/

**Comment by jdubois** [comment]
Indeed! I'm surprised they changed this, but let's follow the Hibernate 6 best practices.
